src/irmin/tree.ml

```diff
@@ -790,16 +788,34 @@ module Make (P : Private.S) = struct
         | `True -> empty_marks ()
         | `Marks n -> n
       in
+      let pre path bindings acc =
+        match pre with
+        | None -> Lwt.return acc
+        | Some pre ->
+            let s = Seq.fold_left (fun acc (s, _) -> s :: acc) [] bindings in
+            pre path s acc
+      in
+      let post path bindings acc =
+        match post with
+        | None -> Lwt.return acc
+        | Some post ->
+            let s = Seq.fold_left (fun acc (s, _) -> s :: acc) [] bindings in
+            post path s acc
+      in
       let rec aux : type r. (t, acc, r) folder =
        fun ~path acc d t k ->
         let apply acc = node path t acc in
         let next acc =
           match force with
-          | `True | `And_clear ->
-              (* XXX: Let's not call [to_map] when [Value] *)
-              let* m = to_map t >|= get_ok "fold" in
-              if force = `And_clear then clear ~depth:0 t;
-              (map [@tailcall]) ~path acc d (Some m) k
+          | `True | `And_clear -> (
+              match t.v with
+              | Map m ->
+                  if force = `And_clear then clear ~depth:0 t;
+                  (map [@tailcall]) ~path acc d (Some m) k
+              | Value (repo, _, _) | Hash (repo, _) ->
+                  let* v = to_value t >|= get_ok "fold" in
+                  if force = `And_clear then clear ~depth:0 t;
+                  (value [@tailcall]) ~path acc d (repo, v) k)
           | `False skip -> (
               match cached_map t with
               | Some n -> (map [@tailcall]) ~path acc d (Some n) k
```

Comment:
maybe rewrite this in a similar fashion to what is done above? We just want to avoid loading new values from the store if `force` is false, but otherwise, the semantics should be very similar to the other case.

Reply:
The only case that was not yet treated here, is when we have a value with an `updatemap`; we iterate over the `Add elt` only now. Not sure if we want this.